source/fundamentals/bson.txt

```diff
@@ -0,0 +1,298 @@
+      - Omit the empty ``LastName`` field
+      - Flatten the nested struct and bring all fields up to the top
+        level
+
+      .. tabs::
+
+   .. tab:: No Struct Tags
+      :tabid: no-struct-tags
+
+      The following example demonstrates how the Go Driver marshals a
+      struct without any struct tags:
+
+      .. code-block:: go
+
+         type Address struct {
+         	Street string
+         	City   string
+         	State  string
+         }
+                  
+         type Student struct {
+         	FirstName string
+         	LastName  string
+         	Address   Address
+         	Age       int
+         }
+         ...
+
+         coll := client.Database("school").Collection("students")
+         address1 := Address{ "1 Lakewood Way", "Elwood City", "PA" }
+         student1 := Student{ FirstName : "Arthur", Address : address1, Age : 8}
+         _, err = coll.InsertOne(context.TODO(), student1)
+
+      The corresponding BSON representation looks like this:
+
+      .. code-block:: json
+         :copyable: false
+      
+         { 
+           "_id" : ObjectId("..."),
+           "firstname" : "Arthur",
+           "lastname" : "",
+           "address": {
+                        "street" : "1 Lakewood Way",
+                        "city" : "Elwood City",
+                        "state" : "PA"
+                      },
+           "age" : 8
+         }
+         
+      Without struct tags, the driver:
+      
+      - Sets the lowercase of the struct fields as the BSON field names
+      - Includes an empty ``lastname`` field
+      - Stores the ``Address`` field as a nested value
+
+Unmarshalling
+-------------
+
+You can unmarshal BSON documents by using the ``Decode()`` method on the
+result of the ``FindOne`` method or any ``*mongo.Cursor`` instance.
+
+The ``Decode()`` method returns an ``error`` type which
+contains one of the following values:
+
+- ``nil`` if a document matched your query, and there were no errors
+  retrieving and unmarshalling the document. 
+- If the driver retrieved your document but could not unmarshal your result, the
+  ``Decode()`` method returns the unmarshalling error.
+- If there was an error retrieving your document during execution of the
+  ``FindOne()`` method, the error propagates to the ``Decode()`` method and
+  the ``Decode()`` method returns the error. 
+
+When used on the ``SingleResult`` type returned by the ``FindOne()``
+method, ``Decode()`` can also return the ``ErrNoDocuments`` error if no
+documents matched the query filter.
+
+The following example demonstrates how you can use the ``Decode()``
+method to unmarshal and read the result of a simple ``FindOne()``
+operation:
+
+.. code-block:: go
+   
+   myCollection := client.Database("school").Collection("students")
+   filter := bson.D{{"age", 8}}
+
+   var result bson.D
+   cursor := myCollection.FindOne(context.TODO(), filter)
+   err = cursor.Decode(&result)
```

Comment:
The type returned by `FindOne` isn't a cursor (it returns a `*mongo.SingleResult`), so the variable naming here is confusing. Combine these two lines into the more common usage of `FindOne`:

```go
err := myCollection.FindOne(context.TODO(), filter).Decode(&result)
```